### ISSUE TITLE
`E0601: used-before-assignment` false positive for import under `TYPE_CHECKING` check

### ISSUE THREAD

**Issue by Rogdham** [PyCQA/pylint#7368]
### Bug description

```python
# pylint: disable=missing-module-docstring, missing-function-docstring

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing_extensions import assert_type
else:

    def assert_type(val, _):
        return val

ANSWER = 42
assert_type(ANSWER, int)
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
************* Module a
a.py:13:0: E0601: Using variable 'assert_type' before assignment (used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 2.86/10 (previous run: 2.86/10, +0.00)
```

### Expected behavior

This is a false positive, as `assert_type` is defined in both branches of the `if`.

### Pylint version

```shell
pylint 2.15.0
astroid 2.12.4
Python 3.10.6 (main, Aug  3 2022, 17:39:45) [GCC 12.1.1 20220730]
```

### OS / Environment

Archlinux

### Additional dependencies

astroid==2.12.4
dill==0.3.5.1
isort==5.10.1
lazy-object-proxy==1.7.1
mccabe==0.7.0
platformdirs==2.5.2
pylint==2.15.0
tomli==2.0.1
tomlkit==0.11.4
wrapt==1.14.1

**Comment by Rogdham** [comment]
The issue is still here if the import is from stdlib instead of an external module:

```python
# pylint: disable=missing-module-docstring, missing-function-docstring

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections import defaultdict
else:
    def defaultdict():
        return {}

defaultdict()
```

```
************* Module a
c.py:11:6: E0601: Using variable 'defaultdict' before assignment (used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 1.67/10 (previous run: 1.67/10, +0.00)
```

**Comment by jacobtylerwalls** [comment]
Good catch. 3159b1744caee14068b4d875bb5adaa5ffbebd44 only checked `AssignName` nodes, not `FunctionDef` nodes, as here. Would you like to prepare a PR?

**Comment by DanielNoord** [comment]
@jacobtylerwalls Just for my understanding, why are we special-casing `TYPE_CHECKING` here?

```python
# pylint: disable=missing-module-docstring, missing-function-docstring

from typing import value

if value:
    from typing_extensions import assert_type
else:

    def assert_type(val, _):
        return val


ANSWER = 42
assert_type(ANSWER, int)
```

This doesn't raise, only a `no-name-in-module` warning.

**Comment by jacobtylerwalls** [comment]
Assignments under if TYPE_CHECKING are always false at runtime, so from the perspective of `used-before-assignment`, we can assume they never happened. With some control-flow knowledge that the name is defined in the `else`, though, we can avoid assuming the name was never assigned. Here the else is only checked for assignments, not function definitions.

**Comment by Rogdham** [comment]
> Would you like to prepare a PR?

@jacobtylerwalls I did what I could in #7370, but the `pylint` codebase is foreign to me so I'm sure I missed tons of things!
